From c0d850f95b735d538ff40857b3b04fbbcdfda84b Mon Sep 17 00:00:00 2001
From: "wenjie.li" <wenjie.li@amlogic.com>
Date: Tue, 28 Oct 2014 15:00:51 +0800
Subject: [PATCH 4994/5965] PD#97549:LCD:add transfer IC anx6345 inital code
 and panel parametr of R140RTN03

Squashed commit of the following:

commit 62a2e7ee3f9a6f9198f9edd1dde9ce9fae4fe806
Author: wenjie.li <wenjie.li@amlogic.com>
Date:   Sat Oct 11 09:51:14 2014 +0800

    PD#97549 add new panel parameter and add ttl transfer edp lane number choice

    Change-Id: Ib21f380adc0084139ba0a8102a1d7d72dec5a092

commit 7bcc6e1e4faace24167306fd6e55a883c05ec9dd
Author: wenjie.li <wenjie.li@amlogic.com>
Date:   Tue Sep 16 15:11:47 2014 +0800

    PD#97549 add transfer IC anx6345 inital code and panel parametr of R140RTN03

    Change-Id: I8f93e6464a0468b536ef123904be260af4390220

Change-Id: I71bb8addc5b25afffbb8be8c3f4cf239b0a8d518
---
 drivers/amlogic/display/vout/aml_lcd.dtd      |  58 ++
 .../amlogic/display/vout/lcd_extern/aml.dtd   |  15 +
 .../display/vout/lcd_extern/edp_anx6345.c     | 581 +++++++++++++++
 .../display/vout/lcd_extern/edp_anx6345.h     | 673 ++++++++++++++++++
 4 files changed, 1327 insertions(+)
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/edp_anx6345.c
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/edp_anx6345.h

diff --git a/drivers/amlogic/display/vout/aml_lcd.dtd b/drivers/amlogic/display/vout/aml_lcd.dtd
index bacc09292f98..e538283932a3 100755
--- a/drivers/amlogic/display/vout/aml_lcd.dtd
+++ b/drivers/amlogic/display/vout/aml_lcd.dtd
@@ -611,6 +611,64 @@
 
         max_lane_count=<1>; /** max lane count supported */
     };
+	
+//$$ MATCH "lcd_model_config_match" = <&lcd_B140RTN02>
+//$$ L2 PROP_STR = "model_name"
+//$$ L2 PROP_STR = "interface"
+//$$ L2 PROP_U32 2 = "active_area"
+//$$ L2 PROP_U32 2 = "lcd_bits_option"
+//$$ L2 PROP_U32 2 = "resolution"
+//$$ L2 PROP_U32 2 = "period"
+//$$ L2 PROP_U32 2 = "clock_hz_pol"
+//$$ L2 PROP_U32 2 = "hsync_width_backporch"
+//$$ L2 PROP_U32 2 = "vsync_width_backporch"
+//$$ L2 PROP_U32 2 = "pol_hsync_vsync"
+//$$ L2 PROP_U32 = "max_lane_count"
+    lcd_B140RTN02:lcd_B140RTN02{
+        model_name="B140RTN02"; /** lcd model name */
+        interface="edp";		/** lcd_interface(mipi, lvds, edp, ttl) */
+        active_area=<309 174>;	/** lcd active_area or display_area horizontal size, vertical size : (unit in mm, you can find them on the home page of lcd spec) */
+        lcd_bits_option=<6 0>;	/** lcd_bits(6,8), bits_option(0=only support one mode as lcd_bits define, 1=both support 6/8bit) */
+
+        resolution=<1600 900>;	/** horizontal resolution, vertical resolution */
+        period=<2032 930>;		/** horizontal period(htotal), vertical period(vtotal) */
+
+        clock_hz_pol=<113385600 0>;	/** clock_hz(unit in Hz, both support clk and frame_rate, >200 regard as clk, <200 regard as frame_rate), clk_pol(only valid for TTL) */
+        hsync_width_backporch=<32 200>;	/** hsync_width, hsync_backporch(include hsync_width) */
+        vsync_width_backporch=<2 23>;	/** vsync_width, vsync_backporch(include vsync_width) */
+        pol_hsync_vsync=<1 1>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
+
+        max_lane_count=<1>; /** max lane count supported */
+    };
+	
+//$$ MATCH "lcd_model_config_match" = <&lcd_LP097QX2>
+//$$ L2 PROP_STR = "model_name"
+//$$ L2 PROP_STR = "interface"
+//$$ L2 PROP_U32 2 = "active_area"
+//$$ L2 PROP_U32 2 = "lcd_bits_option"
+//$$ L2 PROP_U32 2 = "resolution"
+//$$ L2 PROP_U32 2 = "period"
+//$$ L2 PROP_U32 2 = "clock_hz_pol"
+//$$ L2 PROP_U32 2 = "hsync_width_backporch"
+//$$ L2 PROP_U32 2 = "vsync_width_backporch"
+//$$ L2 PROP_U32 2 = "pol_hsync_vsync"
+//$$ L2 PROP_U32 = "max_lane_count"
+	lcd_LP097QX2:lcd_LP097QX2{
+		model_name="LP097QX2";	/** lcd model name */
+		interface="edp";	 /** lcd_interface(mipi, lvds, edp, ttl) */
+		active_area=<163 208>;	/** lcd active_area or display_area horizontal size, vertical size : (unit in mm, you can find them on the home page of lcd spec) */
+		lcd_bits_option=<8 0>;	/** lcd_bits(6,8), bits_option(0=only support one mode as lcd_bits define, 1=both support 6/8bit) */
+
+		resolution=<1536 2048>;	/** horizontal resolution, vertical resolution */
+		period=<1759 2081>;	 /** horizontal period(htotal), vertical period(vtotal) */
+
+		clock_hz_pol=<200000000 0>;	/** clock_hz(unit in Hz, both support clk and frame_rate, >200 regard as clk, <200 regard as frame_rate), clk_pol(only valid for TTL) */
+		hsync_width_backporch=<15 67>;	/** hsync_width, hsync_backporch(include hsync_width) */
+		vsync_width_backporch=<1 4>;	/** vsync_width, vsync_backporch(include vsync_width) */
+		pol_hsync_vsync=<1 1>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
+ 
+		max_lane_count=<4>; /** max lane count supported */
+	};
 
 //******************************************************************************
 //lvds
diff --git a/drivers/amlogic/display/vout/lcd_extern/aml.dtd b/drivers/amlogic/display/vout/lcd_extern/aml.dtd
index ca49f7b104a0..6bd37c0fd73a 100755
--- a/drivers/amlogic/display/vout/lcd_extern/aml.dtd
+++ b/drivers/amlogic/display/vout/lcd_extern/aml.dtd
@@ -44,5 +44,20 @@
 
 		type = <2>;	/** lcd_extern_driver type: 0=i2c, 1=spi, 2=mipi */
 	};
+//$$ DEVICE = "lcd_extern_edp_ANX6345"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "type"
+	lcd_extern_edp_ANX6345{
+		compatible = "amlogic,lcd_anx6345";
+		dev_name ="lcd_anx6345";
+		status = "disabled";
 
+		type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi, 2=mipi */
+		i2c_address = <0x38>; /** i2c address */
+		i2c_bus = "i2c_bus_b";
+		lane_num = <1>; /** edp lane_num:1/2/4 */
+		bits = <0>;  /** bits of lcd(0: 6bit, 1: 8bit) */
+		link_rate = <1>; /** edp link rate (0: 1.62G, 1: 27G, 2: 5.4G) */
+	};	
+	
 #endif
diff --git a/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.c b/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.c
new file mode 100755
index 000000000000..4d80da837690
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.c
@@ -0,0 +1,581 @@
+/*
+ * AMLOGIC lcd external driver.
+ *
+ * Communication protocol:
+ * I2C 
+ *
+ */
+
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+#include <linux/jiffies.h> 
+#include <linux/i2c.h>
+#include <linux/i2c-aml.h>
+#include <linux/miscdevice.h>
+#include <linux/mutex.h>
+#include <linux/mm.h>
+#include <linux/device.h>
+#include <linux/fs.h>
+#include <linux/delay.h>
+#include <linux/sysctl.h>
+#include <asm/uaccess.h>
+#include <mach/pinmux.h>
+#include <mach/gpio.h>
+#include <linux/amlogic/vout/aml_lcd_extern.h>
+#include "edp_anx6345.h"
+
+static struct lcd_extern_config_t *lcd_ext_config = NULL;
+                          
+static struct i2c_client *aml_anx6345_70_client;
+static struct i2c_client *aml_anx6345_72_client;
+
+const char *anx_addr_name[2] = {"anx6345_70","anx6345_72"};
+static unsigned edp_tx_lane = 1;
+
+//#define LCD_EXT_DEBUG_INFO
+#ifdef LCD_EXT_DEBUG_INFO
+#define DBG_PRINT(...)		printk(__VA_ARGS__)
+#else
+#define DBG_PRINT(...)
+#endif
+
+#define LCD_EXTERN_NAME			"lcd_anx6345"
+
+
+static int aml_i2c_write(struct i2c_client *i2client,unsigned char *buff, unsigned len)
+{
+    int res = 0;
+    struct i2c_msg msg[] = {
+        {
+        .addr = i2client->addr,
+        .flags = 0,
+        .len = len,
+        .buf = buff,
+        }
+    };
+    
+    res = i2c_transfer(i2client->adapter, msg, 1);
+    if (res < 0) {
+        printk("%s: i2c transfer failed [addr 0x%02x]\n", __FUNCTION__, i2client->addr);
+    }
+    
+    return res;
+}
+
+static int aml_i2c_read(struct i2c_client *i2client,unsigned char *buff, unsigned len)
+{
+    int res = 0;
+    struct i2c_msg msgs[] = {
+        {
+            .addr = i2client->addr,
+            .flags = 0,
+            .len = 1,
+            .buf = buff,
+        },
+        {
+            .addr = i2client->addr,
+            .flags = I2C_M_RD,
+            .len = len,
+            .buf = buff,
+        }
+    };
+    res = i2c_transfer(i2client->adapter, msgs, 2);
+    if (res < 0) {
+        printk("%s: i2c transfer failed [addr 0x%02x]\n", __FUNCTION__, i2client->addr);
+    }
+
+    return res;
+}
+
+static int i2c_reg_read(unsigned char reg, unsigned char *buf)
+{
+    int ret = 0;
+
+    return ret;
+}
+
+static int i2c_reg_write(unsigned char reg, unsigned char value)
+{
+    int ret=0;
+
+    return ret;
+}
+static int SP_TX_Write_Reg(unsigned char addr,unsigned char reg ,unsigned char data)
+{
+	struct i2c_client *client =aml_anx6345_70_client;
+	unsigned char buff[2];
+	int ret;
+	buff[0] = reg;
+	buff[1] = data;
+	if (addr == 0x70){
+		client=aml_anx6345_70_client;
+	}	
+	else if (addr == 0x72){
+		client=aml_anx6345_72_client;
+	}
+
+	ret =aml_i2c_write(client,	buff, 1);
+	if (ret < 0)
+		return -1;
+	else
+		return 0;
+}
+
+static int SP_TX_Read_Reg(unsigned char addr,unsigned char reg ,unsigned char* data)
+{
+		struct i2c_client *client =aml_anx6345_70_client;
+		int ret;
+		*data = reg;
+
+		if (addr == 0x70){
+			client = aml_anx6345_70_client;
+		}	
+		else if (addr == 0x72){
+			client = aml_anx6345_72_client;
+		}
+					
+		ret = aml_i2c_read(client, data, 1);
+		if (ret < 0)
+			return -1;
+		else
+			return 0;
+}
+
+static int SP_TX_Wait_AUX_Finished(void)
+{
+	unsigned char c;
+	unsigned char cCnt;
+	cCnt = 0;
+
+	SP_TX_Read_Reg(0x70, SP_TX_AUX_STATUS, &c);
+	while(c & 0x10){//aux busy
+		cCnt++;
+		SP_TX_Read_Reg(0x70, SP_TX_AUX_STATUS, &c);
+
+		if(cCnt>100){
+			return 0; //aux fail
+		}
+	}
+
+	return 1;//aux ok
+}
+
+static int SP_TX_AUX_DPCDRead_Bytes(unsigned char addrh, unsigned char addrm, unsigned char addrl,unsigned char cCount,unsigned char *pBuf)
+{
+	unsigned char c,i;
+
+	//clr buffer
+	SP_TX_Write_Reg(0x70, SP_TX_BUF_DATA_COUNT_REG, 0x80);
+
+	//set read cmd and count
+	SP_TX_Write_Reg(0x70, SP_TX_AUX_CTRL_REG, ((cCount-1) <<4)|0x09);
+
+	//set aux address15:0
+	SP_TX_Write_Reg(0x70, SP_TX_AUX_ADDR_7_0_REG, addrl);
+	SP_TX_Write_Reg(0x70, SP_TX_AUX_ADDR_15_8_REG, addrm);
+
+	//set address19:16 and enable aux
+	SP_TX_Read_Reg(0x70, SP_TX_AUX_ADDR_19_16_REG, &c);
+	SP_TX_Write_Reg(0x70, SP_TX_AUX_ADDR_19_16_REG, (c & 0xf0) | addrh);
+
+	//Enable Aux
+	SP_TX_Read_Reg(0x70, SP_TX_AUX_CTRL_REG2, &c);
+	SP_TX_Write_Reg(0x70, SP_TX_AUX_CTRL_REG2, c|0x01);
+
+	mdelay(5);
+	if(!SP_TX_Wait_AUX_Finished())
+	return 0;
+
+	for(i =0;i<cCount;i++) {
+		SP_TX_Read_Reg(0x70, SP_TX_BUF_DATA_0_REG+i, &c);
+
+		*(pBuf+i) = c;
+
+		if(i >= MAX_BUF_CNT)
+		break;
+	}
+
+	return 1;//aux ok
+}
+
+static int lcd_extern_i2c_init(void)
+{
+		unsigned lane_num;
+		unsigned link_rate;
+		unsigned char bits;
+		unsigned char device_id;
+		unsigned char temp;
+		unsigned char temp1;
+		unsigned count =0;
+		unsigned count1 =0;
+		lane_num = edp_tx_lane; //1 lane
+		link_rate = VAL_EDP_TX_LINK_BW_SET_270; //2.7G
+		bits =0; //0x00: 6bit;  0x10:8bit
+		SP_TX_Write_Reg (0x72, 0x05, 0x00);
+		
+		SP_TX_Read_Reg(0x72, 0x01, &device_id);
+		
+		if(device_id==0xaa) {
+			printk("ANX6345 Chip found\n\n\n");
+		}	else {
+			printk("ANX6345 Chip not found\n\n\n");
+			return 1;
+		}
+		temp = device_id;
+		//if aux read fail, do h/w reset,
+		while((!SP_TX_AUX_DPCDRead_Bytes(0x00,0x00,0x00,1,&temp1))&&(count<200)) {
+			//read fail, h/w reset
+			SP_TX_Write_Reg (0x72, 0x06, 0x01);
+			SP_TX_Write_Reg (0x72, 0x06, 0x00);
+			SP_TX_Write_Reg (0x72, 0x05, 0x00);
+			mdelay(10);
+			count++;
+		}
+		
+		//software reset
+		SP_TX_Read_Reg(0x72, SP_TX_RST_CTRL_REG, &temp);
+		SP_TX_Write_Reg(0x72, SP_TX_RST_CTRL_REG, temp | SP_TX_RST_SW_RST);
+		SP_TX_Write_Reg(0x72, SP_TX_RST_CTRL_REG, temp & ~SP_TX_RST_SW_RST);
+		
+		SP_TX_Write_Reg(0x70, SP_TX_EXTRA_ADDR_REG, 0x50);//EDID address for AUX access
+		SP_TX_Write_Reg(0x70, SP_TX_HDCP_CTRL, 0x00);	//disable HDCP polling mode.
+		//SP_TX_Write_Reg(0x70, SP_TX_HDCP_CTRL, 0x02);	//Enable HDCP polling mode.
+		SP_TX_Write_Reg(0x70, SP_TX_LINK_DEBUG_REG, 0x30);//enable M value read out
+		
+		//SP_TX_Read_Reg(0x70, SP_TX_DEBUG_REG1, &temp);
+		SP_TX_Write_Reg(0x70, SP_TX_DEBUG_REG1, 0x00);//disable polling HPD
+		
+		SP_TX_Read_Reg(0x70, SP_TX_HDCP_CONTROL_0_REG, &temp);
+		SP_TX_Write_Reg(0x70, SP_TX_HDCP_CONTROL_0_REG, temp | 0x03);//set KSV valid
+		
+		SP_TX_Read_Reg(0x70, SP_TX_AUX_CTRL_REG2, &temp);
+		SP_TX_Write_Reg(0x70, SP_TX_AUX_CTRL_REG2, temp|0x08);//set double AUX output
+		
+		SP_TX_Write_Reg(0x72, SP_COMMON_INT_MASK1, 0xbf);//unmask pll change int
+		SP_TX_Write_Reg(0x72, SP_COMMON_INT_MASK2, 0xff);//mask all int
+		SP_TX_Write_Reg(0x72, SP_COMMON_INT_MASK3, 0xff);//mask all int
+		SP_TX_Write_Reg(0x72, SP_COMMON_INT_MASK4, 0xff);//mask all int
+		
+		//reset AUX
+		SP_TX_Read_Reg(0x72, SP_TX_RST_CTRL2_REG, &temp);
+		SP_TX_Write_Reg(0x72, SP_TX_RST_CTRL2_REG, temp |SP_TX_AUX_RST);
+		SP_TX_Write_Reg(0x72, SP_TX_RST_CTRL2_REG, temp & (~SP_TX_AUX_RST));
+		
+		//Chip initialization
+		
+		SP_TX_Write_Reg(0x70, SP_TX_SYS_CTRL1_REG, 0x00);
+		mdelay(10);
+		
+			SP_TX_Write_Reg(0x72, SP_TX_VID_CTRL2_REG, bits);
+		
+		//ANX6345 chip analog setting
+		SP_TX_Write_Reg(0x70, SP_TX_PLL_CTRL_REG, 0x00);				  //UPDATE: FROM 0X07 TO 0X00
+		
+		//ANX chip analog setting
+		//	SP_TX_Write_Reg(0x72, ANALOG_DEBUG_REG1, 0x70); 			  //UPDATE: FROM 0XF0 TO 0X70
+		SP_TX_Write_Reg(0x70, SP_TX_LINK_DEBUG_REG, 0x30);
+		
+		//force HPD
+		SP_TX_Write_Reg(0x70, SP_TX_SYS_CTRL3_REG, 0x30);
+		
+		/* enable ssc function */
+		SP_TX_Write_Reg(0x70, 0xA7, 0x00); 		          // disable SSC first
+		SP_TX_Write_Reg(0x70, 0xD0, 0x5f);              // ssc d  0.4%, f0/4 mode
+		SP_TX_Write_Reg(0x70, 0xD1, 0x00);
+		SP_TX_Write_Reg(0x70, 0xD2, 0x75);              // ctrl_th
+		SP_TX_Read_Reg(0x70, 0xA7, &temp);
+		SP_TX_Write_Reg(0x70, 0xA7, temp | 0x10);          // enable SSC
+		SP_TX_Read_Reg(0x72, 0x07, &temp);                 //reset SSC
+		SP_TX_Write_Reg(0x72, 0x07, temp | 0x80);
+		SP_TX_Write_Reg(0x72, 0x07, temp & (~0x80));
+		
+		//Select 2.7G
+		SP_TX_Write_Reg(0x70, SP_TX_LINK_BW_SET_REG, link_rate);		//2.7g:0x0a;1.62g:0x06
+		//Select 2 lanes
+		SP_TX_Write_Reg(0x70, 0xa1, lane_num);
+		
+		SP_TX_Write_Reg(0x70, SP_TX_LINK_TRAINING_CTRL_REG, SP_TX_LINK_TRAINING_CTRL_EN);
+		mdelay(5);
+		SP_TX_Read_Reg(0x70, SP_TX_LINK_TRAINING_CTRL_REG, &temp);
+		/* UPDATE: FROM 0X01 TO 0X80 */
+		while((temp&0x80)!=0) {
+			//debug_puts("Waiting...\n");
+			mdelay(5);
+			count1 ++;
+			if(count1 > 100) {
+				printk("ANX6345 Link training fail\n");
+				break;
+			}
+			SP_TX_Read_Reg(0x70, SP_TX_LINK_TRAINING_CTRL_REG, &temp);
+		}
+		
+
+		SP_TX_Write_Reg(0x72, 0x12, 0x2c);
+		SP_TX_Write_Reg(0x72, 0x13, 0x06);
+		SP_TX_Write_Reg(0x72, 0x14, 0x00);
+		SP_TX_Write_Reg(0x72, 0x15, 0x06);
+		SP_TX_Write_Reg(0x72, 0x16, 0x02);
+		SP_TX_Write_Reg(0x72, 0x17, 0x04);
+		SP_TX_Write_Reg(0x72, 0x18, 0x26);
+		SP_TX_Write_Reg(0x72, 0x19, 0x50);
+		SP_TX_Write_Reg(0x72, 0x1a, 0x04);
+		SP_TX_Write_Reg(0x72, 0x1b, 0x00);
+		SP_TX_Write_Reg(0x72, 0x1c, 0x04);
+		SP_TX_Write_Reg(0x72, 0x1d, 0x18);
+		SP_TX_Write_Reg(0x72, 0x1e, 0x00);
+		SP_TX_Write_Reg(0x72, 0x1f, 0x10);
+		SP_TX_Write_Reg(0x72, 0x20, 0x00);
+		SP_TX_Write_Reg(0x72, 0x21, 0x28);
+		
+		SP_TX_Write_Reg(0x72, 0x11, 0x03);
+	//enable BIST. In normal mode, don't need to config this reg
+	//if want to open BIST,must setting right dat 0x11-0x21 base lcd timing.	
+	//	SP_TX_Write_Reg(0x72, 0x0b, 0x09);//colorbar:08,graystep:09
+		SP_TX_Write_Reg(0x72, 0x08, 0x81);		//SDR:0x81;DDR:0x8f
+		
+		//force HPD and stream valid
+		SP_TX_Write_Reg(0x70, 0x82, 0x33);
+    printk("%s\n", __FUNCTION__);
+    return 0;
+}
+
+static int lcd_extern_i2c_remove(void)
+{
+    int ret=0;
+
+    return ret;
+}
+
+static int lcd_extern_driver_update(void)
+{
+    struct aml_lcd_extern_driver_t* lcd_ext;
+
+    lcd_ext = aml_lcd_extern_get_driver();
+    if (lcd_ext) {
+        lcd_ext->type      = lcd_ext_config->type;
+        lcd_ext->name      = lcd_ext_config->name;
+        lcd_ext->reg_read  = i2c_reg_read;
+        lcd_ext->reg_write = i2c_reg_write;
+        lcd_ext->power_on  = lcd_extern_i2c_init;
+        lcd_ext->power_off = lcd_extern_i2c_remove;
+    }
+    else {
+        printk("[error] %s get lcd_extern_driver failed\n", lcd_ext_config->name);
+    }
+
+    return 0;
+}
+
+static int aml_anx6345_70_probe(struct i2c_client *client, const struct i2c_device_id *id)
+{
+
+    if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C)) {
+        printk("[error] %s: functionality check failed\n", __FUNCTION__);
+        return -ENODEV;
+    }
+    else {
+        aml_anx6345_70_client = client;
+        lcd_extern_driver_update();
+    }
+
+    printk("%s OK\n", __FUNCTION__);
+    return 0;
+}
+static int aml_anx6345_72_probe(struct i2c_client *client, const struct i2c_device_id *id)
+{
+
+    if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C)) {
+        printk("[error] %s: functionality check failed\n", __FUNCTION__);
+        return -ENODEV;
+    }
+    else {
+        aml_anx6345_72_client = client;
+      //  lcd_extern_driver_update();
+    }
+
+    printk("%s OK\n", __FUNCTION__);
+    return 0;
+}
+
+
+static int aml_anx6345_70_remove(struct i2c_client *client)
+{
+    return 0;
+}
+static int aml_anx6345_72_remove(struct i2c_client *client)
+{
+    return 0;
+}
+
+
+static const struct i2c_device_id aml_anx6345_70_id[]= {
+    {"anx6345_70", 0},
+    { }
+};
+static const struct i2c_device_id aml_anx6345_72_id[]= {
+    {"anx6345_72", 0},
+    { }
+};
+
+// MODULE_DEVICE_TABLE(i2c, aml_tc101_id);
+
+static struct i2c_driver aml_anx6345_70_driver= {
+    .probe    = aml_anx6345_70_probe,
+    .remove   = aml_anx6345_70_remove,
+    .id_table = aml_anx6345_70_id,
+    .driver = {
+        .name = "anx6345_70",
+        .owner =THIS_MODULE,
+    },
+};
+static struct i2c_driver aml_anx6345_72_driver= {
+    .probe    = aml_anx6345_72_probe,
+    .remove   = aml_anx6345_72_remove,
+    .id_table = aml_anx6345_72_id,
+    .driver = {
+        .name = "anx6345_72",
+        .owner =THIS_MODULE,
+    },
+};
+
+
+static int aml_anx6345_probe(struct platform_device *pdev)
+{
+    struct i2c_board_info i2c_info[2];
+    struct i2c_adapter *adapter;
+    struct i2c_client *i2c_client;
+   // struct lcd_extern_config_t extern_config[3];
+    int i = 0;
+    int ret = 0;
+
+    if (lcd_extern_driver_check()) {
+        return -1;
+    }
+    if (lcd_ext_config == NULL)
+        lcd_ext_config = kzalloc(sizeof(*lcd_ext_config), GFP_KERNEL);
+    if (lcd_ext_config == NULL) {
+        printk("[error] %s probe: failed to alloc data\n", LCD_EXTERN_NAME);
+        return -1;
+    }
+
+    pdev->dev.platform_data = lcd_ext_config;
+
+    if (get_lcd_extern_dt_data(pdev->dev.of_node, lcd_ext_config) != 0) {
+        printk("[error] %s probe: failed to get dt data\n", LCD_EXTERN_NAME);
+        goto lcd_extern_probe_failed;
+    }
+   for(i=0; i<2; i++){
+    	memset(&i2c_info[i], 0, sizeof(i2c_info[i]));
+   }
+
+    adapter = i2c_get_adapter(lcd_ext_config->i2c_bus);
+    if (!adapter) {
+        printk("[error] %s£ºfailed to get i2c adapter\n", LCD_EXTERN_NAME);
+        goto lcd_extern_probe_failed;
+    }
+    for(i=0; i<2; i++){
+    	strncpy(i2c_info[i].type, anx_addr_name[i], I2C_NAME_SIZE);
+    }
+    i2c_info[0].addr = 0x38;//0x70 >> 1;
+    i2c_info[1].addr = 0x39;//0x72 >>1;
+    for(i=0; i<2; i++){
+    	//i2c_info[i].platform_data = extern_config[i];
+    	i2c_info[i].flags = 0;
+    	if(i2c_info[i].addr > 0x7f)
+      		i2c_info[i].flags = 0x10;
+    	i2c_client = i2c_new_device(adapter, &i2c_info[i]);
+			if (!i2c_client) {
+					printk("[error] %x :failed to new i2c device\n", i2c_info[i].addr);
+					goto lcd_extern_probe_failed;
+			}
+			else{
+					DBG_PRINT("[error] %x: new i2c device succeed\n",i2c_info[i].addr);
+			}
+   }
+    
+
+    if (!aml_anx6345_70_client) {
+        ret = i2c_add_driver(&aml_anx6345_70_driver);
+        if (ret) {
+            printk("[error] lcd_extern probe: add i2c_driver failed\n");
+            goto lcd_extern_probe_failed;
+        }
+    }
+    
+    if (!aml_anx6345_72_client) {
+        ret = i2c_add_driver(&aml_anx6345_72_driver);
+        if (ret) {
+            printk("[error] lcd_extern probe: add i2c_driver failed\n");
+            goto lcd_extern_probe_failed;
+        }
+    }
+    
+
+    printk("%s ok\n", __FUNCTION__);
+    return ret;
+
+lcd_extern_probe_failed:
+    if (lcd_ext_config)
+        kfree(lcd_ext_config);
+    return -1;
+}
+
+static int aml_anx6345_remove(struct platform_device *pdev)
+{
+    remove_lcd_extern(lcd_ext_config);
+    if (pdev->dev.platform_data)
+        kfree (pdev->dev.platform_data);
+    return 0;
+}
+
+#ifdef CONFIG_USE_OF
+static const struct of_device_id aml_anx6345_dt_match[]={
+    {
+        .compatible = "amlogic,lcd_anx6345",
+    },
+    {},
+};
+#else
+#define aml_tc101_dt_match NULL
+#endif
+
+static struct platform_driver aml_anx6345_driver = {
+    .probe  = aml_anx6345_probe,
+    .remove = aml_anx6345_remove,
+    .driver = {
+        .name  = LCD_EXTERN_NAME,
+        .owner = THIS_MODULE,
+#ifdef CONFIG_USE_OF
+        .of_match_table = aml_anx6345_dt_match,
+#endif
+    },
+};
+
+static int __init aml_anx6345_init(void)
+{
+    int ret;
+    DBG_PRINT("%s\n", __FUNCTION__);
+
+    ret = platform_driver_register(&aml_anx6345_driver);
+    if (ret) {
+        printk("[error] %s failed to register lcd extern driver module\n", __FUNCTION__);
+        return -ENODEV;
+    }
+    return ret;
+}
+
+static void __exit aml_anx6345_exit(void)
+{
+    platform_driver_unregister(&aml_anx6345_driver);
+}
+
+//late_initcall(aml_tc101_init);
+module_init(aml_anx6345_init);
+module_exit(aml_anx6345_exit);
+
+MODULE_AUTHOR("AMLOGIC");
+MODULE_DESCRIPTION("LCD Extern driver for TC101");
+MODULE_LICENSE("GPL");
diff --git a/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.h b/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.h
new file mode 100755
index 000000000000..d8cb084b2ff1
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/edp_anx6345.h
@@ -0,0 +1,673 @@
+#ifndef EDP_ANX6345_H
+#define EDP_ANX6345_H
+
+
+#define SP_TX_PORT0_ADDR 0x70
+#define SP_TX_PORT1_ADDR 0x74
+#define SP_TX_PORT2_ADDR 0x72
+#define MIPI_RX_PORT1_ADDR 0x7A
+
+#define MAX_BUF_CNT 6
+
+#define CR_LOOP_TIME 5
+#define EQ_LOOP_TIME 5
+
+#define SP_TX_AVI_SEL 0x01
+#define SP_TX_SPD_SEL 0x02
+#define SP_TX_MPEG_SEL 0x04
+
+#define VAL_EDP_TX_LINK_BW_SET_162						0x06
+#define VAL_EDP_TX_LINK_BW_SET_270						0x0a
+#define VAL_EDP_TX_LINK_BW_SET_540						0x14
+
+//End for DEV_addr 0x7A/0x7E
+
+/***************************************************************/
+//  DEV_ADDR = 0x70 or 0x78 , Displayport mode and HDCP registers
+#define SP_TX_HDCP_STATUS							  				0x00
+#define SP_TX_HDCP_AUTH_PASS						  			0x02//bit position
+
+#define SP_TX_HDCP_CONTROL_0_REG                  		0x01
+#define SP_TX_HDCP_CONTROL_0_STORE_AN            0x80//bit position
+#define SP_TX_HDCP_CONTROL_0_RX_REPEATER   	0x40//bit position
+#define SP_TX_HDCP_CONTROL_0_RE_AUTH              0x20//bit position
+#define SP_TX_HDCP_CONTROL_0_SW_AUTH_OK       0x10//bit position
+#define SP_TX_HDCP_CONTROL_0_HARD_AUTH_EN   0x08//bit position
+#define SP_TX_HDCP_CONTROL_0_HDCP_ENC_EN      0x04//bit position
+#define SP_TX_HDCP_CONTROL_0_BKSV_SRM_PASS  0x02//bit position
+#define SP_TX_HDCP_CONTROL_0_KSVLIST_VLD        0x01//bit position
+
+
+#define SP_TX_HDCP_CONTROL_1_REG                  		0x02
+#define SP_TX_HDCP_CONTROL_1_DDC_NO_STOP      			0x20//bit position
+#define SP_TX_HDCP_CONTROL_1_DDC_NO_ACK        			0x10//bit position
+#define SP_TX_HDCP_CONTROL_1_EDDC_NO_ACK          		0x08//bit position
+//#define SP_TX_HDCP_CONTROL_1_HDCP_EMB_SCREEN_EN   		0x04//bit position
+#define SP_TX_HDCP_CONTROL_1_RCV_11_EN                  0x02//bit position
+#define SP_TX_HDCP_CONTROL_1_HDCP_11_EN           		0x01//bit position
+
+#define SP_TX_HDCP_LINK_CHK_FRAME_NUM				 	0x03
+#define SP_TX_HDCP_CONTROL_2_REG						0x04
+
+#define SP_TX_HDCP_AKSV0								0x05
+#define SP_TX_HDCP_AKSV1								0x06
+#define SP_TX_HDCP_AKSV2								0x07
+#define SP_TX_HDCP_AKSV3								0x08
+#define SP_TX_HDCP_AKSV4								0x09
+
+//AKSV
+#define SP_TX_HDCP_AN0									0x0A
+#define SP_TX_HDCP_AN1									0x0B
+#define SP_TX_HDCP_AN2									0x0C
+#define SP_TX_HDCP_AN3									0x0D
+#define SP_TX_HDCP_AN4									0x0E
+#define SP_TX_HDCP_AN5									0x0F
+#define SP_TX_HDCP_AN6									0x10
+#define SP_TX_HDCP_AN7									0x11
+
+//BKSV
+#define SP_TX_HDCP_BKSV0								0x12
+#define SP_TX_HDCP_BKSV1								0x13
+#define SP_TX_HDCP_BKSV2								0x14
+#define SP_TX_HDCP_BKSV3								0x15
+#define SP_TX_HDCP_BKSV4								0x16
+
+#define SP_TX_HDCP_R0_L									0x17
+#define SP_TX_HDCP_R0_H									0x18
+
+#ifndef M_VID_0
+#define M_VID_0 0xC0
+#define M_VID_1 0xC1
+#define M_VID_2 0xC2
+#define N_VID_0 0xC3
+#define N_VID_1 0xC4
+#define N_VID_2 0xC5
+#endif
+
+#define SP_TX_HDCP_R0_WAIT_Timer					 0x40
+
+
+
+#define SP_TX_SYS_CTRL1_REG           					0x80
+//#define SP_TX_SYS_CTRL1_PD_IO         					0x80    // bit position
+//#define SP_TX_SYS_CTRL1_PD_VID        					0x40    // bit position
+//#define SP_TX_SYS_CTRL1_PD_LINK       					0x20    // bit position
+//#define SP_TX_SYS_CTRL1_PD_TOTAL      					0x10    // bit position
+//#define SP_TX_SYS_CTRL1_MODE_SEL      					0x08    // bit position
+#define SP_TX_SYS_CTRL1_DET_STA       					0x04    // bit position
+#define SP_TX_SYS_CTRL1_FORCE_DET     					0x02    // bit position
+#define SP_TX_SYS_CTRL1_DET_CTRL      					0x01    // bit position
+
+#define SP_TX_SYS_CTRL2_REG           					0x81
+// #define SP_TX_SYS_CTRL2_ENHANCED 	  					0x08	  //bit position
+#define SP_TX_SYS_CTRL2_CHA_STA       					0x04    // bit position
+#define SP_TX_SYS_CTRL2_FORCE_CHA     					0x02    // bit position
+#define SP_TX_SYS_CTRL2_CHA_CTRL      					0x01    // bit position
+
+#define SP_TX_SYS_CTRL3_REG           					0x82
+#define SP_TX_SYS_CTRL3_HPD_STATUS    					0x40    // bit position
+#define SP_TX_SYS_CTRL3_F_HPD         					0x20    // bit position
+#define SP_TX_SYS_CTRL3_HPD_CTRL      					0x10    // bit position
+#define SP_TX_SYS_CTRL3_STRM_VALID    					0x04    // bit position
+#define SP_TX_SYS_CTRL3_F_VALID       					0x02    // bit position
+#define SP_TX_SYS_CTRL3_VALID_CTRL    					0x01    // bit position
+
+#define SP_TX_SYS_CTRL4_REG			  					0x83
+#define SP_TX_SYS_CTRL4_ENHANCED 	  					0x08//bit position
+
+#define SP_TX_VID_CTRL				  					0x84
+
+#define SP_TX_AUD_CTRL									0x87
+#define SP_TX_AUD_CTRL_AUD_EN							0x01
+
+
+#define SP_TX_PKT_EN_REG              					0x90
+#define SP_TX_PKT_AUD_UP								0x80  // bit position
+#define SP_TX_PKT_AVI_UD              					0x40  // bit position
+#define SP_TX_PKT_MPEG_UD             					0x20  // bit position    
+#define SP_TX_PKT_SPD_UD              					0x10  // bit position   
+#define SP_TX_PKT_AUD_EN								0x08  // bit position=
+#define SP_TX_PKT_AVI_EN              					0x04  // bit position          
+#define SP_TX_PKT_MPEG_EN             					0x02  // bit position     
+#define SP_TX_PKT_SPD_EN              					0x01  // bit position       
+
+
+#define SP_TX_HDCP_CTRL 												0x92
+
+#define SP_TX_LINK_BW_SET_REG         				 0xA0
+#define SP_TX_LANE_COUNT_SET_REG      				 0xA1
+
+#define SP_TX_TRAINING_PTN_SET_REG                   0xA2
+#define SP_TX_SCRAMBLE_DISABLE						 0x20//bit 5
+
+#define SP_TX_TRAINING_LANE0_SET_REG                 				0xA3
+#define SP_TX_TRAINING_LANE0_SET_MAX_PRE_REACH        0x20        // bit position
+#define SP_TX_TRAINING_LANE0_SET_MAX_DRIVE_REACH     0x04        // bit position
+
+#define SP_TX_TRAINING_LANE1_SET_REG                0xA4
+
+
+#define SSC_CTRL_REG1					 0xA7
+#define SPREAD_AMP						 0x10//bit 4
+#define MODULATION_FREQ					 0x01//bit 0
+
+
+#define SP_TX_LINK_TRAINING_CTRL_REG                0xA8
+#define SP_TX_LINK_TRAINING_CTRL_EN                 0x01        // bit position
+
+
+#define SP_TX_DEBUG_REG1							0xB0
+#define SP_TX_DEBUG_HPD_POLLING_DET						0x40//bit position
+#define SP_TX_DEBUG_HPD_POLLING_EN						0x20//bit position
+#define SP_TX_DEBUG_PLL_LOCK						0x10//bit position
+
+
+#define SP_TX_LINK_DEBUG_REG                        0xB8
+#define SP_TX_LINK_DEBUG_INSERT_ER                  0x02        // bit position
+#define SP_TX_LINK_DEBUG_PRBS31_EN                  0x01        // bit position
+
+#define SP_TX_SINK_COUNT_REG                0xB9
+
+#define SP_TX_LINK_STATUS_REG1                               0xBB
+
+#define SP_TX_SINK_STATUS_REG                                   0xBE
+#define SP_TX_SINK_STATUS_SINK_STATUS_1          	0x02        // bit position
+#define SP_TX_SINK_STATUS_SINK_STATUS_0          	0x01        // bit position
+
+
+//#define SP_TX_LINK_TEST_COUNT                     0xC0
+
+
+#define SP_TX_PLL_CTRL_REG											0xC7	
+#define SP_TX_PLL_CTRL_PLL_PD           						0x80        // bit position
+#define SP_TX_PLL_CTRL_PLL_RESET        					0x40        // bit position 
+//#define SP_TX_PLL_CTRL_CPREG_BLEED      					0x08        // bit position 
+
+#define SP_TX_ANALOG_POWER_DOWN_REG                   			0xC8
+#define SP_TX_ANALOG_POWER_DOWN_MACRO_PD              	0x20        // bit position 
+#define SP_TX_ANALOG_POWER_DOWN_AUX_PD                		0x10        // bit position 
+//#define SP_TX_ANALOG_POWER_DOWN_CH3_PD                		0x08        // bit position 
+//#define SP_TX_ANALOG_POWER_DOWN_CH2_PD                		0x04        // bit position 
+#define SP_TX_ANALOG_POWER_DOWN_CH1_PD                		0x02        // bit position 
+#define SP_TX_ANALOG_POWER_DOWN_CH0_PD                		0x01        // bit position 
+
+
+#define SP_TX_ANALOG_TEST_REG                         					0xC9
+#define SP_TX_ANALOG_TEST_MACRO_RST                   				0x20       // bit position 
+#define SP_TX_ANALOG_TEST_PLL_TEST                    				0x10       // bit position 
+#define SP_TX_ANALOG_TEST_CH3_TEST                    				0x08       // bit position 
+#define SP_TX_ANALOG_TEST_CH2_TEST                    				0x04       // bit position 
+#define SP_TX_ANALOG_TEST_CH1_TEST                    				0x02       // bit position 
+#define SP_TX_ANALOG_TEST_CH0_TEST                    				0x01       // bit position 
+
+#define SP_TX_GNS_CTRL_REG                            							0xCD
+#define SP_EQ_LOOP_CNT											0x40//bit position
+#define SP_TX_VIDEO_MAP_CTRL                 			                            0x02       // bit position 
+#define SP_TX_RS_CTRL                        					              	0x01       // bit position 
+
+#define SP_TX_DOWN_SPREADING_CTRL1                                               0xD0   //guochuncheng
+#define SP_TX_DOWN_SPREADING_CTRL2                                               0xD1
+#define SP_TX_DOWN_SPREADING_CTRL3                                               0xD2
+#define SP_TX_SSC_D_CTRL                                                             0x40       //bit position
+#define SP_TX_FS_CTRL_TH_CTRL                                                   0x20       //bit position
+
+#define SP_TX_M_CALCU_CTRL												0xD9
+#define M_GEN_CLK_SEL													0x01//bit 0
+
+
+#define SP_TX_EXTRA_ADDR_REG											0xCE
+#define SP_TX_I2C_STRETCH_CTRL_REG                                                              0xDB
+#define SP_TX_AUX_STATUS            										0xE0
+#define SP_TX_DEFER_CTRL_REG            									0xE2
+#define SP_TXL_DEFER_CTRL_EN  					                     		       0x80       // bit position 
+
+#define SP_TX_BUF_DATA_COUNT_REG											0xE4
+#define SP_TX_AUX_CTRL_REG              										0xE5
+#define SP_TX_MOT_BIT													0x04//bit 2
+
+#define SP_TX_AUX_ADDR_7_0_REG          									0xE6
+#define SP_TX_AUX_ADDR_15_8_REG         									0xE7
+#define SP_TX_AUX_ADDR_19_16_REG        									0xE8
+
+#define SP_TX_AUX_CTRL_REG2                                                 0xE9
+#define SP_TX_ADDR_ONLY_BIT													0x02//bit 1
+
+#define SP_TX_BUF_DATA_0_REG                          0xf0
+#define SP_TX_BUF_DATA_1_REG                          0xf1
+#define SP_TX_BUF_DATA_2_REG                          0xf2
+#define SP_TX_BUF_DATA_3_REG                          0xf3
+#define SP_TX_BUF_DATA_4_REG                          0xf4
+#define SP_TX_BUF_DATA_5_REG                          0xf5
+#define SP_TX_BUF_DATA_6_REG                          0xf6
+#define SP_TX_BUF_DATA_7_REG                          0xf7
+#define SP_TX_BUF_DATA_8_REG                          0xf8
+#define SP_TX_BUF_DATA_9_REG                          0xf9
+#define SP_TX_BUF_DATA_10_REG                         0xfa
+#define SP_TX_BUF_DATA_11_REG                         0xfb
+#define SP_TX_BUF_DATA_12_REG                         0xfc
+#define SP_TX_BUF_DATA_13_REG                         0xfd
+#define SP_TX_BUF_DATA_14_REG                         0xfe
+#define SP_TX_BUF_DATA_15_REG                         0xff
+
+//End for Address 0x70 or 0x78
+
+/***************************************************************/
+//  DEV_ADDR = 0x72 or 0x76, System control registers
+#define SP_TX_VND_IDL_REG             	0x00
+#define SP_TX_VND_IDH_REG             	0x01
+#define SP_TX_DEV_IDL_REG             	0x02
+#define SP_TX_DEV_IDH_REG             	0x03
+#define SP_TX_DEV_REV_REG             	0x04
+
+#define SP_POWERD_CTRL_REG			  	0x05
+#define SP_POWERD_REGISTER_REG			0x80// bit position
+//#define SP_POWERD_MISC_REG			  	0x40// bit position
+#define SP_POWERD_IO_REG			  	0x20// bit position
+#define SP_POWERD_AUDIO_REG				0x10// bit position
+#define SP_POWERD_VIDEO_REG			  	0x08// bit position
+#define SP_POWERD_LINK_REG			  	0x04// bit position
+#define SP_POWERD_TOTAL_REG			  	0x02// bit position
+#define SP_MODE_SEL_REG				  	0x01// bit position
+
+#define SP_TX_RST_CTRL_REG            	0x06
+#define SP_TX_RST_MISC_REG 			  	0x80	// bit position
+#define SP_TX_RST_VIDCAP_REG		  	0x40	// bit position
+#define SP_TX_RST_VIDFIF_REG          	0x20    // bit position
+#define SP_TX_RST_AUDFIF_REG          	0x10    // bit position
+#define SP_TX_RST_AUDCAP_REG         	0x08    // bit position
+#define SP_TX_RST_HDCP_REG            	0x04    // bit position
+#define SP_TX_RST_SW_RST             	0x02    // bit position
+#define SP_TX_RST_HW_RST             	0x01    // bit position
+
+#define SP_TX_RST_CTRL2_REG				0x07
+#define SP_TX_RST_SSC					0x80//bit position
+#define SP_TX_AC_MODE					0x40//bit position
+//#define SP_TX_DDC_RST					0x10//bit position
+//#define SP_TX_TMDS_BIST_RST				0x08//bit position
+#define SP_TX_AUX_RST					0x04//bit position
+#define SP_TX_SERDES_FIFO_RST			0x02//bit position
+#define SP_TX_I2C_REG_RST				0x01//bit position
+
+
+#define SP_TX_VID_CTRL1_REG           	0x08
+#define SP_TX_VID_CTRL1_VID_EN       0x80    // bit position
+#define SP_TX_VID_CTRL1_VID_MUTE   0x40    // bit position
+#define SP_TX_VID_CTRL1_DE_GEN      0x20    // bit position
+#define SP_TX_VID_CTRL1_DEMUX        0x10    // bit position
+#define SP_TX_VID_CTRL1_IN_BIT		  	0x04    // bit position
+#define SP_TX_VID_CTRL1_DDRCTRL		0x02    // bit position
+#define SP_TX_VID_CTRL1_EDGE		  		0x01    // bit position
+
+#define SP_TX_VID_CTRL2_REG           	0x09
+#define SP_TX_VID_CTRL1_YCBIT_SEL  		0x04    // bit position
+
+#define SP_TX_VID_CTRL3_REG           	0x0A
+
+#define SP_TX_VID_CTRL4_REG           	0x0B
+#define SP_TX_VID_CTRL4_E_SYNC_EN	  	0x80	  //bit position
+#define SP_TX_VID_CTRL4_EX_E_SYNC    0x40    // bit position
+#define SP_TX_VID_CTRL4_BIST          		0x08    // bit position
+#define SP_TX_VID_CTRL4_BIST_WIDTH   0x04        // bit position
+
+#define SP_TX_VID_CTRL5_REG           		0x0C
+
+#define SP_TX_VID_CTRL6_REG           		0x0D
+#define SP_TX_VID_UPSAMPLE					0x02//bit position
+
+#define SP_TX_VID_CTRL7_REG           		0x0E
+#define SP_TX_VID_CTRL8_REG           		0x0F
+#define SP_TX_VID_CTRL9_REG           		0x10
+
+#define SP_TX_VID_CTRL10_REG           	0x11
+#define SP_TX_VID_CTRL10_INV_F         	0x08    // bit position
+#define SP_TX_VID_CTRL10_I_SCAN        	0x04    // bit position
+#define SP_TX_VID_CTRL10_VSYNC_POL   0x02    // bit position
+#define SP_TX_VID_CTRL10_HSYNC_POL   0x01    // bit position
+
+#define SP_TX_TOTAL_LINEL_REG         0x12
+#define SP_TX_TOTAL_LINEH_REG         0x13
+#define SP_TX_ACT_LINEL_REG           0x14
+#define SP_TX_ACT_LINEH_REG           0x15
+#define SP_TX_VF_PORCH_REG            0x16
+#define SP_TX_VSYNC_CFG_REG           0x17
+#define SP_TX_VB_PORCH_REG            0x18
+#define SP_TX_TOTAL_PIXELL_REG        0x19
+#define SP_TX_TOTAL_PIXELH_REG        0x1A
+#define SP_TX_ACT_PIXELL_REG          0x1B
+#define SP_TX_ACT_PIXELH_REG          0x1C
+#define SP_TX_HF_PORCHL_REG           0x1D
+#define SP_TX_HF_PORCHH_REG           0x1E
+#define SP_TX_HSYNC_CFGL_REG          0x1F
+#define SP_TX_HSYNC_CFGH_REG          0x20
+#define SP_TX_HB_PORCHL_REG           0x21
+#define SP_TX_HB_PORCHH_REG           0x22
+
+#define SP_TX_VID_STATUS						0x23
+
+#define SP_TX_TOTAL_LINE_STA_L        0x24
+#define SP_TX_TOTAL_LINE_STA_H        0x25
+#define SP_TX_ACT_LINE_STA_L          0x26
+#define SP_TX_ACT_LINE_STA_H          0x27
+#define SP_TX_V_F_PORCH_STA           0x28
+#define SP_TX_V_SYNC_STA              0x29
+#define SP_TX_V_B_PORCH_STA           0x2A
+#define SP_TX_TOTAL_PIXEL_STA_L       0x2B
+#define SP_TX_TOTAL_PIXEL_STA_H       0x2C
+#define SP_TX_ACT_PIXEL_STA_L         0x2D
+#define SP_TX_ACT_PIXEL_STA_H         0x2E
+#define SP_TX_H_F_PORCH_STA_L         0x2F
+#define SP_TX_H_F_PORCH_STA_H         0x30
+#define SP_TX_H_SYNC_STA_L            0x31
+#define SP_TX_H_SYNC_STA_H            0x32
+#define SP_TX_H_B_PORCH_STA_L         0x33
+#define SP_TX_H_B_PORCH_STA_H         0x34
+
+#define SP_TX_Video_Interface_BIST    0x35
+
+#define SPDIF_AUDIO_CTRL0			0x36
+#define SPDIF_AUDIO_CTRL0_SPDIF_IN  0x80 // bit position
+
+#define SPDIF_AUDIO_STATUS0			0x38
+#define SPDIF_AUDIO_STATUS0_CLK_DET 0x80
+#define SPDIF_AUDIO_STATUS0_AUD_DET 0x01
+
+#define SPDIF_AUDIO_STATUS1 0x39
+
+#define AUDIO_BIST_CTRL 0x3c
+#define AUDIO_BIST_EN 0x01
+
+//#define AUDIO_BIST_CHANNEL_STATUS1 0xd0
+//#define AUDIO_BIST_CHANNEL_STATUS2 0xd1
+//#define AUDIO_BIST_CHANNEL_STATUS3 0xd2
+//#define AUDIO_BIST_CHANNEL_STATUS4 0xd3
+//#define AUDIO_BIST_CHANNEL_STATUS5 0xd4
+
+#define SP_TX_VIDEO_BIT_CTRL_0_REG                    0x40
+#define SP_TX_VIDEO_BIT_CTRL_1_REG                    0x41
+#define SP_TX_VIDEO_BIT_CTRL_2_REG                    0x42
+#define SP_TX_VIDEO_BIT_CTRL_3_REG                    0x43
+#define SP_TX_VIDEO_BIT_CTRL_4_REG                    0x44
+#define SP_TX_VIDEO_BIT_CTRL_5_REG                    0x45
+#define SP_TX_VIDEO_BIT_CTRL_6_REG                    0x46
+#define SP_TX_VIDEO_BIT_CTRL_7_REG                    0x47
+#define SP_TX_VIDEO_BIT_CTRL_8_REG                    0x48
+#define SP_TX_VIDEO_BIT_CTRL_9_REG                    0x49
+#define SP_TX_VIDEO_BIT_CTRL_10_REG                   0x4a
+#define SP_TX_VIDEO_BIT_CTRL_11_REG                   0x4b
+#define SP_TX_VIDEO_BIT_CTRL_12_REG                   0x4c
+#define SP_TX_VIDEO_BIT_CTRL_13_REG                   0x4d
+#define SP_TX_VIDEO_BIT_CTRL_14_REG                   0x4e
+#define SP_TX_VIDEO_BIT_CTRL_15_REG                   0x4f
+#define SP_TX_VIDEO_BIT_CTRL_16_REG                   0x50
+#define SP_TX_VIDEO_BIT_CTRL_17_REG                   0x51
+#define SP_TX_VIDEO_BIT_CTRL_18_REG                   0x52
+#define SP_TX_VIDEO_BIT_CTRL_19_REG                   0x53
+#define SP_TX_VIDEO_BIT_CTRL_20_REG                   0x54
+#define SP_TX_VIDEO_BIT_CTRL_21_REG                   0x55
+#define SP_TX_VIDEO_BIT_CTRL_22_REG                   0x56
+#define SP_TX_VIDEO_BIT_CTRL_23_REG                   0x57
+#define SP_TX_VIDEO_BIT_CTRL_24_REG                   0x58
+#define SP_TX_VIDEO_BIT_CTRL_25_REG                   0x59
+#define SP_TX_VIDEO_BIT_CTRL_26_REG                   0x5a
+#define SP_TX_VIDEO_BIT_CTRL_27_REG                   0x5b
+#define SP_TX_VIDEO_BIT_CTRL_28_REG                   0x5c
+#define SP_TX_VIDEO_BIT_CTRL_29_REG                   0x5d
+#define SP_TX_VIDEO_BIT_CTRL_30_REG                   0x5e
+#define SP_TX_VIDEO_BIT_CTRL_31_REG                   0x5f
+#define SP_TX_VIDEO_BIT_CTRL_32_REG                   0x60
+#define SP_TX_VIDEO_BIT_CTRL_33_REG                   0x61
+#define SP_TX_VIDEO_BIT_CTRL_34_REG                   0x62
+#define SP_TX_VIDEO_BIT_CTRL_35_REG                   0x63
+#define SP_TX_VIDEO_BIT_CTRL_36_REG                   0x64
+#define SP_TX_VIDEO_BIT_CTRL_37_REG                   0x65
+#define SP_TX_VIDEO_BIT_CTRL_38_REG                   0x66
+#define SP_TX_VIDEO_BIT_CTRL_39_REG                   0x67
+#define SP_TX_VIDEO_BIT_CTRL_40_REG                   0x68
+#define SP_TX_VIDEO_BIT_CTRL_41_REG                   0x69
+#define SP_TX_VIDEO_BIT_CTRL_42_REG                   0x6a
+#define SP_TX_VIDEO_BIT_CTRL_43_REG                   0x6b
+#define SP_TX_VIDEO_BIT_CTRL_44_REG                   0x6c
+#define SP_TX_VIDEO_BIT_CTRL_45_REG                   0x6d
+#define SP_TX_VIDEO_BIT_CTRL_46_REG                   0x6e
+#define SP_TX_VIDEO_BIT_CTRL_47_REG                   0x6f
+
+//AVI info frame
+#define SP_TX_AVI_TYPE              0x70
+#define SP_TX_AVI_VER               0x71
+#define SP_TX_AVI_LEN               0x72
+#define SP_TX_AVI_DB0		     0x73
+#define SP_TX_AVI_DB1               0x74
+#define SP_TX_AVI_DB2               0x75
+#define SP_TX_AVI_DB3               0x76
+#define SP_TX_AVI_DB4               0x77
+#define SP_TX_AVI_DB5               0x78
+#define SP_TX_AVI_DB6               0x79
+#define SP_TX_AVI_DB7               0x7A
+#define SP_TX_AVI_DB8               0x7B
+#define SP_TX_AVI_DB9               0x7C
+#define SP_TX_AVI_DB10              0x7D
+#define SP_TX_AVI_DB11              0x7E
+#define SP_TX_AVI_DB12              0x7F
+#define SP_TX_AVI_DB13              0x80
+#define SP_TX_AVI_DB14              0x81
+#define SP_TX_AVI_DB15              0x82
+
+//Audio info frame
+#define SP_TX_AUD_TYPE			 0x83
+#define SP_TX_AUD_VER			 0x84
+#define SP_TX_AUD_LEN			 0x85
+#define SP_TX_AUD_DB0			 0x86
+#define SP_TX_AUD_DB1			 0x87
+#define SP_TX_AUD_DB2			 0x88
+#define SP_TX_AUD_DB3			 0x89
+#define SP_TX_AUD_DB4			 0x8A
+#define SP_TX_AUD_DB5			 0x8B
+#define SP_TX_AUD_DB6			 0x8C
+#define SP_TX_AUD_DB7			 0x8D
+#define SP_TX_AUD_DB8			 0x8E
+#define SP_TX_AUD_DB9			 0x8F
+#define SP_TX_AUD_DB10			 0x90
+
+//SPD info frame
+#define SP_TX_SPD_TYPE                0x91
+#define SP_TX_SPD_VER                 0x92
+#define SP_TX_SPD_LEN                 0x93
+#define SP_TX_SPD_DATA0		0x94
+#define SP_TX_SPD_DATA1               0x95
+#define SP_TX_SPD_DATA2               0x96
+#define SP_TX_SPD_DATA3               0x97
+#define SP_TX_SPD_DATA4               0x98
+#define SP_TX_SPD_DATA5               0x99
+#define SP_TX_SPD_DATA6               0x9A
+#define SP_TX_SPD_DATA7               0x9B
+#define SP_TX_SPD_DATA8               0x9C
+#define SP_TX_SPD_DATA9               0x9D
+#define SP_TX_SPD_DATA10              0x9E
+#define SP_TX_SPD_DATA11              0x9F
+#define SP_TX_SPD_DATA12              0xA0
+#define SP_TX_SPD_DATA13              0xA1
+#define SP_TX_SPD_DATA14              0xA2
+#define SP_TX_SPD_DATA15              0xA3
+#define SP_TX_SPD_DATA16              0xA4
+#define SP_TX_SPD_DATA17              0xA5
+#define SP_TX_SPD_DATA18              0xA6
+#define SP_TX_SPD_DATA19              0xA7
+#define SP_TX_SPD_DATA20              0xA8
+#define SP_TX_SPD_DATA21              0xA9
+#define SP_TX_SPD_DATA22              0xAA
+#define SP_TX_SPD_DATA23              0xAB
+#define SP_TX_SPD_DATA24              0xAC
+#define SP_TX_SPD_DATA25              0xAD
+#define SP_TX_SPD_DATA26              0xAE
+#define SP_TX_SPD_DATA27              0xAF
+
+//Mpeg source info frame
+#define SP_TX_MPEG_TYPE               0xB0
+#define SP_TX_MPEG_VER                0xB1
+#define SP_TX_MPEG_LEN                0xB2
+#define SP_TX_MPEG_DATA0              0xB3
+#define SP_TX_MPEG_DATA1              0xB4
+#define SP_TX_MPEG_DATA2              0xB5
+#define SP_TX_MPEG_DATA3              0xB6
+#define SP_TX_MPEG_DATA4              0xB7
+#define SP_TX_MPEG_DATA5              0xB8
+#define SP_TX_MPEG_DATA6              0xB9
+#define SP_TX_MPEG_DATA7              0xBA
+#define SP_TX_MPEG_DATA8              0xBB
+#define SP_TX_MPEG_DATA9              0xBC
+#define SP_TX_MPEG_DATA10             0xBD
+#define SP_TX_MPEG_DATA11            0xBE
+#define SP_TX_MPEG_DATA12            0xBF
+#define SP_TX_MPEG_DATA13            0xC0
+#define SP_TX_MPEG_DATA14            0xC1
+#define SP_TX_MPEG_DATA15            0xC2
+#define SP_TX_MPEG_DATA16            0xC3
+#define SP_TX_MPEG_DATA17            0xC4
+#define SP_TX_MPEG_DATA18            0xC5
+#define SP_TX_MPEG_DATA19            0xC6
+#define SP_TX_MPEG_DATA20            0xC7
+#define SP_TX_MPEG_DATA21            0xC8
+#define SP_TX_MPEG_DATA22            0xC9
+#define SP_TX_MPEG_DATA23            0xCA
+#define SP_TX_MPEG_DATA24            0xCB
+#define SP_TX_MPEG_DATA25            0xCC
+#define SP_TX_MPEG_DATA26            0xCD
+#define SP_TX_MPEG_DATA27            0xCE
+
+//#define GNSS_CTRL_REG				0xCD
+//#define ENABLE_SSC_FILTER			0x80//bit 
+
+//#define SSC_D_VALUE					 0xD0
+//#define SSC_CTRL_REG2					 0xD1
+
+#define ANALOG_DEBUG_REG1			0xDC
+#define ANALOG_SEL_BG				0x40//bit 4
+#define ANALOG_SWING_A_30PER		0x08//bit 3
+
+#define ANALOG_DEBUG_REG2			0xDD
+#define ANALOG_24M_SEL				0x08//bit 3
+//#define ANALOG_FILTER_ENABLED		0x10//bit 4
+
+
+#define ANALOG_DEBUG_REG3			0xDE
+
+#define PLL_FILTER_CTRL1			0xDF
+#define PD_RING_OSC					0x40//bit 6
+
+#define PLL_FILTER_CTRL2			0xE0
+#define PLL_FILTER_CTRL3			0xE1
+#define PLL_FILTER_CTRL4			0xE2
+#define PLL_FILTER_CTRL5			0xE3
+#define PLL_FILTER_CTRL6			0xE4
+
+#define SP_TX_I2S_CTRL			0xE6
+#define SP_TX_I2S_FMT			0xE7
+#define SP_TX_I2S_CH_Status1			0xD0
+#define SP_TX_I2S_CH_Status2			0xD1
+#define SP_TX_I2S_CH_Status3			0xD2
+#define SP_TX_I2S_CH_Status4			0xD3
+#define SP_TX_I2S_CH_Status5			0xD4
+
+//interrupt
+#define SP_COMMON_INT_STATUS1     0xF1
+#define SP_COMMON_INT1_PLL_LOCK_CHG 	0x40//bit position
+#define SP_COMMON_INT1_VIDEO_FORMAT_CHG 0x08//bit position
+#define SP_COMMON_INT1_AUDIO_CLK_CHG	0x04//bit position
+#define SP_COMMON_INT1_VIDEO_CLOCK_CHG  0x02//bit position
+
+
+#define SP_COMMON_INT_STATUS2	  0xF2
+#define SP_COMMON_INT2_AUTHCHG	  0x02 //bit position
+#define SP_COMMON_INT2_AUTHDONE	  0x01 //bit position
+
+#define SP_COMMON_INT_STATUS3	  0xF3
+#define SP_COMMON_INT3_AFIFO_UNDER	0x80//bit position
+#define SP_COMMON_INT3_AFIFO_OVER	0x40//bit position
+
+#define SP_COMMON_INT_STATUS4	    0xF4
+#define SP_COMMON_INT4_PLUG                0x01   // bit position
+#define SP_COMMON_INT4_ESYNC_ERR          0x10   // bit position
+#define SP_COMMON_INT4_HPDLOST		0x02   //bit position
+#define SP_COMMON_INT4_HPD_CHANGE   0x04   //bit position
+
+
+#define SP_TX_INT_STATUS1		  0xF7
+#define SP_TX_INT_STATUS1_HPD	  0x40 //bit position
+#define SP_TX_INT_STATUS1_TRAINING_Finish       0x20   // bit position
+#define SP_TX_INT_STATUS1_POLLING_ERR        0x10   // bit position
+
+#define SP_TX_INT_SINK_CHG		  0x08//bit position
+
+//interrupt mask
+#define SP_COMMON_INT_MASK1			  0xF8
+#define SP_COMMON_INT_MASK2			  0xF9
+#define SP_COMMON_INT_MASK3			  0xFA
+#define SP_COMMON_INT_MASK4			  0xFB
+#define SP_INT_MASK					  					0xFE
+#define SP_TX_INT_CTRL_REG            		0xFF
+//End for dev_addr 0x72 or 0x76
+
+/***************************************************************/
+/***************************************************************/
+
+//DPCD regs
+#define DPCD_DPCD_REV                                                0x00
+#define DPCD_MAX_LINK_RATE                                      0x01
+#define DPCD_MAX_LANE_COUNT                                   0x02
+#define DPCD_MAX_DOWNSPREAD                                  0x03
+#define DPCD_NORP                                                        0x04
+#define DPCD_DOWNSTREAMPORT_PRESENT                  0x05
+
+#define DPCD_RECEIVE_PORT0_CAP_0                        0x08
+#define DPCD_RECEIVE_PORT0_CAP_1                        0x09
+#define DPCD_RECEIVE_PORT0_CAP_2                        0x0a
+#define DPCD_RECEIVE_PORT0_CAP_3                        0x0b
+
+#define DPCD_LINK_BW_SET                                      0x00
+#define DPCD_LANE_COUNT_SET                                0x01
+#define DPCD_TRAINING_PATTERN_SET                     0x02
+#define DPCD_TRAINNIG_LANE0_SET                         0x03
+#define DPCD_TRAINNIG_LANE1_SET                         0x04
+#define DPCD_TRAINNIG_LANE2_SET                         0x05
+#define DPCD_TRAINNIG_LANE3_SET                         0x06
+#define DPCD_DOWNSPREAD_CTRL                            0x07
+
+#define DPCD_SINK_COUNT                                          0x00
+#define DPCD_DEVICE_SERVICE_IRQ_VECTOR              0x01
+#define DPCD_LANE0_1_STATUS                                   0x02
+#define DPCD_LANE2_3_STATUS                                   0x03
+#define DPCD_LANE_ALIGN_STATUS_UPDATED            0x04
+#define DPCD_SINK_STATUS                                          0x05
+#define DPCD_ADJUST_REQUEST_LANE0_1                     0x06
+#define DPCD_ADJUST_REQUEST_LANE2_3                     0x07
+#define DPCD_TRAINING_SCORE_LANE0                       0x08
+#define DPCD_TRAINING_SCORE_LANE1                       0x09
+#define DPCD_TRAINING_SCORE_LANE2                       0x0a
+#define DPCD_TRAINING_SCORE_LANE3                       0x0b
+
+#define DPCD_TEST_REQUEST                               0x18
+#define DPCD_TEST_LINK_RATE                             0x19
+
+#define DPCD_TEST_LANE_COUNT                            0x20
+
+#define DPCD_TEST_Response                              0x60
+#define TEST_ACK                                                  0x01
+#define DPCD_TEST_EDID_Checksum_Write                   0x04//bit position
+
+#define DPCD_TEST_EDID_Checksum                         0x61
+
+
+#define DPCD_SPECIFIC_INTERRUPT                               0x10
+#define DPCD_USER_COMM1                                             0x22//define for downstream HDMI Rx sense detection
+
+
+
+
+
+
+
+
+
+
+
+#endif
\ No newline at end of file
-- 
2.19.0

